### ISSUE TITLE
clang-analyzer-core checks are very slow

### ISSUE THREAD

**Issue by carlosgalvezp** [llvm/llvm-project#59747]
Hi!

As I posted in this [other issue](https://github.com/llvm/llvm-project/issues/59642#issuecomment-1363223257), I notice that `clang-analyzer-*` (in particular `clang-analyzer-core*`) checks are extremely slow (I am running them through `clang-tidy`).

For comparison, Clang compiler takes less than 2 seconds to analyze the TU, while `clang-analyzer-core` checks take almost 6 minutes to go through it. That's 180x slower.

Before spending time preparing a minimal example out of my original code, I wanted to ask some preliminary questions:

* Clang analyzer is supposed to be slower than compilation, but is it normal that it is **that much** slower? 
* Is there any particular setting that makes the checks slower? Can I configure the tool to trade off speed vs accuracy? I cannot remove individual checks, because clang-tidy will enable all or nothing.
* How can I profile the checks further? 

Thanks!

**Comment by haoNoQ** [comment]
Yeah that's a bit too slow. The typical ratio is 2-3x, around 5x is probably fine, I've seen 10x on some relatively normal code, but anything above that definitely sounds like an exotic cornercase.

The rough idea is that theoretically speaking, the kind of analysis conducted by clang analyzer can consume indefinite amount of time if allowed to. The reason why it finishes at all is because it has a number of limited budgets it isn't allowed to exceed, such as the total size of the exploration graph, or number of nested function calls to descend into. (And we're pretty smart in extracting the most value out of those budgets.) If certain code takes a lot of time to analyze, that's because it manages to stay within *all* these limits for so long. So it's likely that the solution on our side is going to be to add more limits. Sometimes there are real performance bugs though. So one way or another, I'm very interested in taking a look.

You can profile it yourself a little bit. In order to do that, first obtain a `clang --analyze` invocation (usually by simply adding `--analyze` to the translation unit's normal compilation command, eg. from the compilation database), and then add `-Xclang -analyzer-display-progress`. It'll show you time spent on every "entry point" of the analysis - i.e. every function that gets analyzed outside of any caller context, but the analysis may descend into its *callees* and spend time there, so the culprit is either that function or whatever it calls. This will hopefully help you pinpoint the code that takes a long time to analyze. Another nice flag is `-Xclang -analyzer-stats` which will list most budgets at end of invocation, but that's probably best left for us to decypher :D

There are a few flags to control budgets, which may help you as a workaround, but I don't think they're exposed through `clang-tidy` either. The most common one is `Xclang -analyzer-config -Xclang mode=shallow` which is a preset that'll lower most budgets and ruin most interprocedural analysis, resulting in analysis time that's pretty close to compilation time. There are also flags for controlling individual budgets, but we don't really recommend that, because results could be quite counter-intuitive, so it's better to stick to either deep or shallow mode.

**Comment by haoNoQ** [comment]
So if it's an option for you to attach a preprocessed reproducer (`clang -E`) that's the best way to report the bug. We can reduce it ourselves. If this isn't an option (eg. your code is proprietary), then I guess the only way to report the bug would be for you to simulate the problem with a toy example.

**Comment by carlosgalvezp** [comment]
Thanks a lot for the quick feedback! I'll try out those configuration options. I might be able to pass those options to `clang-tidy` since you can pass any regular compilation arguments to it. 

Yes, unfortunately the code is proprietary so I'll need to clean out the proprietary things from it. Will definitely send a preprocessed source for you guys to take a look :)

**Comment by carlosgalvezp** [comment]
So thanks to your commands I got some new info. The time is spent here:

```
ANALYZE (Path,  Inline_Regular): foo.cpp MyTest<std::tuple<std::pair<std::integral_constant<unsigned long, 4>, std::integral_constant<unsigned long, 4>>, std::integral_constant<unsigned long, 7>>>::TestBody() : 3802.4 ms
```

Roughly, the TU is a GoogleTest that has just one test function `MyTest`, but it's a _typed test_ instantiated for ~80 types. So the analyzer always reports around 5 seconds for each `MyTest` instantiation.

The final statistics look like:

```
===-------------------------------------------------------------------------===
                                Analyzer timers
===-------------------------------------------------------------------------===
  Total Execution Time: 313.3927 seconds (313.5445 wall clock)

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
  308.1420 (100.0%)   5.2463 (100.0%)  313.3884 (100.0%)  313.5401 (100.0%)  Path exploration time
   0.0033 (  0.0%)   0.0009 (  0.0%)   0.0042 (  0.0%)   0.0043 (  0.0%)  Syntax-based analysis time
   0.0002 (  0.0%)   0.0000 (  0.0%)   0.0002 (  0.0%)   0.0002 (  0.0%)  Path-sensitive report post-processing time
  308.1455 (100.0%)   5.2473 (100.0%)  313.3927 (100.0%)  313.5445 (100.0%)  Total
```

Unfortunately `-analyzer-stats` didn't seem to show anything, or at least I didn't see any difference. What's it supposed to look like?

I can certainly refactor the test and move some logic from compile time to runtime, but I'm still curious how clang (and other clang-tidy checks) can deal with it so fast and yet clang-analyzer takes so long. Will post a minimal example as soon as I get it :)

**Comment by carlosgalvezp** [comment]
Here is a reduced minimal example that reproduces the problem on a smaller scale:

https://godbolt.org/z/WYW99Gcvv

This takes around 1 second to run with Clang (`-fsyntax-only`) and around 5 seconds to run clang analyzer. The time taken by clang analyzer increases linearly with the number of types used in the test, whereas the Clang time is unaffected by this (whether I use -fsyntax-only or not). From the logs, we can clearly see that clang-analyzer is analyzing each instantiation so each new type adds time.

Hope that it helps, let me know if I can provide anything else! Looking forward to your findings :)
